src/lib.rs

```diff
@@ -432,4 +432,51 @@ mod test {
 
         Ok(())
     }
+
+    fn null() -> Result<File> {
+        File::create(if cfg!(windows) {
+            "nul"
+        } else if cfg!(target_os = "redox") {
+            "null:"
+        } else {
+            "/dev/null"
+        })
+    }
+
+    #[test]
+    fn test_multiple() -> Result<()> {
+        let null = null()?;
+
+        let guard_1 = StdoutOverride::from_io_ref(&null)?;
+        let guard_2 = StdoutOverride::from_io_ref(&null)?;
+
+        std::mem::forget(guard_2);
+        drop(guard_1);
+
+        Ok(())
+    }
+
+    #[test]
+    fn test_multiple_panic() -> Result<()> {
+        let null = null()?;
+
+        let guard_0 = StdoutOverride::from_io_ref(&null)?;
+        let guard_1 = StdoutOverride::from_io_ref(&null)?;
+        let guard_2 = StdoutOverride::from_io_ref(&null)?;
+        drop(guard_1);
```

Comment:
Shouldn't this panic?
is there some clever reasoning I'm not seeing? because in my mind it sounds like you either need to drop them in the same order, or panic on every drop. I can't quite see how you can recover if 1 dropped out of order, I think?
@Kestrer

Reply:
When `guard_2` is dropped, it will panic but it won't change the standard output, so it will be the same as it was before `guard_1` was created.

In other words, when `guard_1` is dropped we skip dropping `guard_2` and assume the user has called `std::mem::forget` or `Box::leak` or created a reference cycle with it. This assumption is only violated when `guard_2` is actually dropped - but in that case we will keep stdout the same, so we don't set stdout to above the end of the "stdout stack".

Or, graphically:

```
Start:
| native stdout |

guard_0 is created
| null stdout (guard_0) |
| native stdout |

guard_1 is created
| null stdout (guard_1) |
| null stdout (guard_0) |
| native stdout |

guard_2 is created
| null stdout (guard_2) |
| null stdout (guard_1) |
| null stdout (guard_0) |
| native stdout |

guard_1 is dropped
| null stdout (guard_0) |
| native stdout |

guard_2 is dropped - Panic!
| null stdout (guard_0) |
| native stdout |

guard_0 is dropped
| native stdout |
```

Reply:
hmm that's an interesting approach,
Thanks for explaining!